### PR TITLE
support pages offline

### PR DESCRIPTION
**Pull request by simcha-waintrob**

### CODE REVIEW

File: cpi-side/addon-cpi.ts [hunk 1]
```diff
@@ -1,13 +1,40 @@
...
+        const pageKey = req.params.pageKey;
+        if (pageKey) {
+            const service = new ClientPagesService();
+            result = await service.getPageData(pageKey);
+        }
+    } catch (err) {
+        console.log(err);
+        result = {
+            error: JSON.stringify(err)
+        }
```
Comment (Chasky-H) [inline cpi-side/addon-cpi.ts:33]: see the documentation on how to handle errors

File: cpi-side/client-pages-service.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    async prepageLocalPage(page: Page): Promise<Page> {
+        debugger
+        // through each slideshow block, replace the image url with the correct url
+        // TODO in the future, this, shouldn't be hardcoded!
+        await Promise.all(page.Blocks.map((block: any) => {
+            const configuration = block.Configuration;
+            if (configuration.Resource === 'Slideshow') {
+                configuration.Data.slides.forEach(async (slide: any) => {
+                    const assetKey = slide.image.asset;
+                    const assetUrl = await pepperi["pfs"].getFile(assetKey);
```
Comment (Chasky-H) [inline cpi-side/client-pages-service.ts:36]: how will you know to get it from the assets pfs table?

File: cpi-side/client-pages-service.ts [hunk 3]
```diff
@@ -21,21 +21,27 @@ class ClientPagesService {
...
         // TODO in the future, this, shouldn't be hardcoded!
-        await Promise.all(page.Blocks.map((block: any) => {
+        await Promise.all(page.Blocks.map(async (block: any) => {
             const configuration = block.Configuration;
             if (configuration.Resource === 'Slideshow') {
-                configuration.Data.slides.forEach(async (slide: any) => {
+                await Promise.all(configuration.Data.slides.map(async (slide: any) => {
                     const assetKey = slide.image.asset;
-                    const assetUrl = await pepperi["pfs"].getFile(assetKey);
+                    const assetUrl = (await pepperi.files.assets.get(assetKey)).URL
```
Comment (Chasky-H) [inline cpi-side/client-pages-service.ts:35]: null check. 
If it returns undefined leave the online URL as is

File: cpi-side/files-service.ts [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                resolve();
+            }).catch(err => {
+                console.log(err);
+                reject(err);
+            });
+        });        
+    }
+
+    getFlatFilesList(blocks: BlockFiles[]): BlockFile[] {
+        let flatFilesList: BlockFile[] = [];
```
Comment (Chasky-H) [inline cpi-side/files-service.ts:123]: try: 
```typescript
blocks.map(x => (x.Files && Array.isArray(x.Files)) ? x.Files : []).flat()
```

**Review by Chasky-H — APPROVED** [review]

**Comment by TomerPopik** [comment]
Great

**Pull request closed without merging**